src/functional/otf/compilation/build_systems/compiledb.py

```diff
@@ -0,0 +1,293 @@
+                    entry["command"],
+                    cwd=self.root_path,
+                    shell=True,
+                    stdout=log_file_pointer,
+                    stderr=log_file_pointer,
+                )
+
+        build_data.update_status(new_status=build_data.BuildStatus.COMPILED, path=self.root_path)
+
+    def run_config(self):
+        compile_db = json.loads(self.compile_commands_cache.read_text())
+
+        (self.root_path / "build").mkdir(exist_ok=True)
+        (self.root_path / "bin").mkdir(exist_ok=True)
+
+        for entry in compile_db:
+            for key, value in entry.items():
+                entry[key] = (
+                    value.replace("$NAME", self.fencil_name)
+                    .replace("$BINDINGS_FILE", self.bindings_file_name)
+                    .replace("$SRC_PATH", str(self.root_path))
+                )
+
+        (self.root_path / "compile_commands.json").write_text(json.dumps(compile_db))
+
+        build_data.write_data(
+            build_data.BuildData(
+                status=build_data.BuildStatus.CONFIGURED,
+                module=pathlib.Path(compile_db[-1]["output"]),
+                entry_point_name=self.fencil_name,
+            ),
+            self.root_path,
+        )
+
+
+def _cc_cache_name(
+    deps: tuple[source.LibraryDependency, ...], build_type: str, flags: list[str]
+) -> str:
+    fencil_name = "compile_commands_cache"
+    deps_str = "_".join(f"{dep.name}_{dep.version}" for dep in deps)
+    flags_str = "_".join(flags)
+    return "_".join([fencil_name, deps_str, build_type, flags_str]).replace(".", "_")
+
+
+def _cc_cache_module(
+    deps: tuple[source.LibraryDependency, ...],
+    build_type: str,
+    cmake_flags: list[str],
+) -> stages.ProgramSource:
+    name = _cc_cache_name(deps, build_type, cmake_flags)
+    return stages.ProgramSource(
+        entry_point=source.Function(name=name, parameters=()),
+        source_code="",
+        library_deps=deps,
+        language=languages.Cpp,
+        language_settings=languages.LanguageWithHeaderFilesSettings(
+            formatter_key="",
+            formatter_style=None,
+            file_extension="",
+            header_extension="",
+        ),
+    )
+
+
+def _cc_get_compiledb(
+    source_module: stages.ProgramSource, cache_strategy: cache.Strategy
+) -> Optional[pathlib.Path]:
+    cache_path = cache.get_cache_folder(
+        stages.CompilableSource(source_module, None), cache_strategy
+    )
+    compile_db_path = cache_path / "compile_commands.json"
+    if compile_db_path.exists():
+        return compile_db_path
+    return None
+
+
+def _cc_generate_compiledb(
+    source_module: stages.ProgramSource,
+    build_type: str,
+    cmake_flags: list[str],
+    cache_strategy: cache.Strategy,
+) -> pathlib.Path:
+    name = source_module.entry_point.name
+    cache_path = cache.get_cache_folder(
+        stages.CompilableSource(source_module, None), cache_strategy
+    )
```

Comment:
I find this a bit of an abuse of the cache system. You create a dummy program source with barely any data just so that you can use the existing cache system to acquire a folder. The cache folder for the compile DB prototypes don't need to follow the same scheme, they merely have to be in the same session or persistent cache root folder. You could directly use the barely any data to form the subfolders instead of abusing the existing system. This would require duplicating or opening up parts of the existing cache system, which also has its drawbacks.

Overall, I think it's fine if we keep it like this, because the other method might not be cleaner. I added suggestions to make this behaviour explicit by renaming related function to advertise that they are creating only *prototype* programs.

Reply:
As you noted, I was trying to stick to existing APIs where possible given the width of this PR. The caching API has only one public function.